sbt-crossproject/src/main/scala/sbtcrossproject/CrossProject.scala

```diff
@@ -234,30 +236,69 @@ object CrossProject {
         }
       }
 
-      Seq(
-        unmanagedSourceDirectories in Compile ++= {
-          makeCrossSources(crossType.sharedSrcDir(baseDirectory.value, "main"),
-                           scalaBinaryVersion.value,
-                           crossPaths.value)
-        },
-        unmanagedSourceDirectories in Test ++= {
-          makeCrossSources(crossType.sharedSrcDir(baseDirectory.value, "test"),
-                           scalaBinaryVersion.value,
-                           crossPaths.value)
+      def makeSharedSettings(key: SettingKey[Seq[File]], config: String): Map[Platform, Seq[Setting[_]]] = {
+        val partiallyShared = makePartiallySharedSettings(platforms) { platformSubset =>
+          Seq(
+            key ++= makeCrossSources(crossType.partiallySharedSrcDir(baseDirectory.value, platformSubset, config), scalaBinaryVersion.value, crossPaths.value)
+          )
         }
-      )
+        
+        val shared = Seq(
+          key ++= makeCrossSources(crossType.sharedSrcDir(baseDirectory.value, config), scalaBinaryVersion.value, crossPaths.value)
+        )
+
+        partiallyShared.mapValues(_ ++ shared) // add shared to each platform
+      }
+
+      val compileSettings = makeSharedSettings(unmanagedSourceDirectories in Compile, "main")
+      val testSettings = makeSharedSettings(unmanagedSourceDirectories in Test, "test")
+      mergeMapSeq(compileSettings, testSettings)
     }
 
-    private def sharedResourcesSettings(
-        crossType: CrossType): Seq[Setting[_]] = {
-      Seq(
-        unmanagedResourceDirectories in Compile ++= {
-          crossType.sharedResourcesDir(baseDirectory.value, "main")
-        },
-        unmanagedResourceDirectories in Test ++= {
-          crossType.sharedResourcesDir(baseDirectory.value, "test")
+    private def sharedResourcesSettings(crossType: CrossType): Map[Platform, Seq[Setting[_]]] = {
+      def makeSharedSettings(key: SettingKey[Seq[File]], config: String): Map[Platform, Seq[Setting[_]]] = {
+        val partiallyShared = makePartiallySharedSettings(platforms) { platformSubset =>
+          Seq(
+            key ++= crossType.partiallySharedResourcesDir(baseDirectory.value, platformSubset, config).toSeq
+          )
         }
-      )
+        
+        val shared = Seq(
+          key ++= crossType.sharedResourcesDir(baseDirectory.value, config).toSeq
+        )
+
+        partiallyShared.mapValues(_ ++ shared) // add shared to each platform
+      }
+
+      val compileSettings = makeSharedSettings(unmanagedResourceDirectories in Compile, "main")
+      val testSettings = makeSharedSettings(unmanagedResourceDirectories in Test, "test")
+      mergeMapSeq(compileSettings, testSettings)
+    }
+
+    private def makePartiallySharedSettings(platforms: Seq[Platform])(mkSettings: Seq[Platform] => Seq[Setting[_]]): Map[Platform, Seq[Setting[_]]] = {
+      platforms.to[ListSet] // preserve the platform ordering
```

Comment:
I suppose another option here is to use `SortedSet` and impose a lexical ordering instead. This might be less surprising for users.

Reply:
I would definitely recommend using lexicographic ordering, indeed. `CrossProject` is not really designed with the idea that the platforms are *ordered* in any way, so relying on that order is brittle.

Reply:
Done, thanks.